src/ripple/basics/CompressionAlgorithms.h

```diff
@@ -0,0 +1,233 @@
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2020 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLED_COMPRESSIONALGORITHMS_H_INCLUDED
+#define RIPPLED_COMPRESSIONALGORITHMS_H_INCLUDED
+
+#include <ripple/basics/varint_common.h>
+#include <ripple/basics/contract.h>
+#include <lz4frame.h>
+#include <array>
+#include <cstring>
+
+namespace ripple {
+
+namespace compression_algorithms {
+
+inline void do_throw(const char *message)
+{
+    Throw<std::runtime_error>(message);
+}
+
+template<typename BufferFactory>
+std::pair<void const*, std::size_t>
+lz4f_compress(void const * in,
+              std::size_t in_size, BufferFactory&& bf)
+{
+    std::pair<void const*, std::size_t> result;
+    std::array<std::uint8_t, varint_traits<std::uint32_t>::max> vi;
+
+    if (in_size > UINT32_MAX)
+        do_throw("lz4f compress: invalid size");
+
+    auto const original_size_bytes = write_varint(vi.data(), in_size);
+
+    std::size_t const out_capacity = LZ4F_compressFrameBound(in_size, NULL);
+
+    auto* compressed = bf(original_size_bytes + out_capacity);
+    result.first = compressed;
+
+    std::memcpy(compressed, vi.data(), original_size_bytes);
+
+    size_t compressed_size = LZ4F_compressFrame(
+            compressed + original_size_bytes,
+            out_capacity,
+            in,
+            in_size,
+            NULL);
+    if (LZ4F_isError(compressed_size))
+        do_throw("lz4f failed compress update");
+
+    result.second = original_size_bytes + compressed_size;
+
+    return result;
+}
+
+template<typename InputStream>
+bool
+copy_stream(InputStream& in, std::uint8_t* dst, int size, std::vector<int> &used_size)
+{
+    const void * data = NULL;
+    int data_size = 0;
+    int copied = 0;
+
+    while (copied != size && in.Next(&data, &data_size))
+    {
+        auto sz = data_size >= (size - copied) ? (size - copied) : data_size;
+        std::memcpy(dst + copied, data, sz);
+        copied += sz;
+        used_size.push_back(data_size);
+    }
+
+    return copied == size;
+}
+
+template<typename InputStream>
+std::size_t
+get_original_size(InputStream& in)
```

Comment:
It looks good. The only thing is that it makes assumptions about `read_varint` implementation (`val & 0x80`).  What if we overload `read_varint`: `read_varint(void const* buf, std::size_t buflen, std::size_t& n, std::size_t& t)`
It takes extra argument `n` which is the number of bytes read so far from the `buf`. `buf` always points to the start of `varint` header. The function returns -1 if the `buf` is too small. Then `getOriginalSize` is something like:

    template<typename InputStream>
    std::size_t
    getOriginalSize1(InputStream& in)
    {
        constexpr size_t maxSize = varint_traits<std::uint32_t>::max;
        boost::container::static_vector<std::uint8_t, maxSize> vi;
        std::size_t originalSize = 0;
        std::uint8_t const* data = nullptr;
        int dataSize = 0;
        std::size_t used = 0;
        int originalSizeBytes = 0;
        std::size_t chunksBytes = 0;

        while (in.Next(reinterpret_cast<void const**>(&data), &dataSize)) {

            if (dataSize == 0)
                continue;

            auto size = dataSize > (maxSize - vi.size()) ? (maxSize - vi.size()) : dataSize;

            if (!vi.empty())
            {
                // save bytes consumed so far from the stream - need to know how many bytes to back up
                // from the last data chunk
                chunksBytes = vi.size();
                vi.insert(vi.begin() + vi.size(), data, data + size);
                data = vi.data();
                size = vi.size();
            }

            originalSizeBytes = read_varint(data, size, used, originalSize);

            if (originalSizeBytes == 0)
                doThrow("lz4f decompress:: originalSizeBytes == 0");
            else if (originalSizeBytes > 0)
                break;

            if (vi.empty())
                vi.insert(vi.begin(), data, data + size);
        }

        if (dataSize > (originalSizeBytes - chunksBytes))
            in.BackUp(dataSize - (originalSizeBytes - chunksBytes));

        return originalSize;
    }
If the first chunk is large enough then `read_varint` parses the header in first iteration. This is the vast majority of cases. Otherwise the chunks are copied into the `vi`. We only `BackUp` the last chunk.

Reply:
Yup, I think that looks good. I don't have time to take a close look now, but go ahead and push it and I'll look at it closely when I hit a good stopping point.

Reply:
Thanks. Will do.

Reply:
On closer look, the fact that `read_varint` now returns a signed type worries me. If the result is mistaken assigned to an `size_t` (as easy mistake to make, I think), then the code compiles without warning, but does the wrong thing. (left a comment in the code too).